### PR TITLE
Move exception handling to ParallelUtility

### PR DESCRIPTION
**Pull request by yufeih**
#5729 

Makes it easier to write parallel looks without try catch.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5747)

### CODE REVIEW

**Comment by yufeih** [comment]
@OsmondJiang , do you know why `azure-docs-pr` has these validation diffs? Does some one changed the ruleset?

**Comment by yufeih** [comment]
`devsandbox` diff is due to `errorCode` fix. The repo has build error.

**Comment by OsmondJiang** [comment]
> @OsmondJiang , do you know why `azure-docs-pr` has these validation diffs? Did some one changed the ruleset?

@yufeih today we synced the latest allow list data from PROD to PPE per Yanjun's requests, seems this sync brings new values for ms.service.

image

Add @kfpwolf.

File: src/docfx/lib/log/ErrorLog.cs [hunk 1]
```diff
@@ -35,7 +35,7 @@ internal sealed class ErrorLog : IDisposable
 
         public IEnumerable<FilePath> ErrorFiles => _errorFiles;
 
-        public ErrorLog(string? outputPath, bool legacy = false)
+        public ErrorLog(string? outputPath = null, bool legacy = false)
```
Comment (live1206) [inline src/docfx/lib/log/ErrorLog.cs:38]: if `outputPath = null` is only for testing, can we remove the default value?

File: src/docfx/lib/log/ErrorLog.cs [hunk 2]
```diff
@@ -240,16 +239,11 @@ private bool ExceedMaxErrors(Config? config, ErrorLevel level)
...
-            {
-                return false;
-            }
-
             return level switch
             {
-                ErrorLevel.Error => Interlocked.Increment(ref _errorCount) > config.MaxErrors,
-                ErrorLevel.Warning => Interlocked.Increment(ref _warningCount) > config.MaxWarnings,
-                ErrorLevel.Suggestion => Interlocked.Increment(ref _suggestionCount) > config.MaxSuggestions,
+                ErrorLevel.Error => Interlocked.Increment(ref _errorCount) > (config?.MaxErrors ?? int.MaxValue),
```
Comment (live1206) [inline src/docfx/lib/log/ErrorLog.cs:244]: why not return earlier?

File: src/docfx/lib/log/ErrorLog.cs (hunk 1 above)
Comment (yufeih) [reply src/docfx/lib/log/ErrorLog.cs:38]: sure

File: src/docfx/lib/log/ErrorLog.cs (hunk 2 above)
Comment (yufeih) [reply src/docfx/lib/log/ErrorLog.cs:244]: `_errorCount` would be incorrect.

**Comment by live1206** [comment]
> `devsandbox` diff is due to `errorCode` fix. The repo has build error.
> 
> @live1206 , do you think the diff is expected?

I am just wondering, maybe we should not return 0 from docs-pipeline when publish is skipped. The build should just fail when there is build error for `en-us` repo

**Comment by mingwli** [comment]
> Why 'base_path' diff in DevSandbox?

It's just a line change.

**Pull request merged**